modules/dxp/apps/analytics/analytics-reports-web/src/main/resources/META-INF/resources/js/context/ChartStateContext.js

```diff
@@ -9,44 +9,134 @@
 
+const INITIAL_STATE = {
+	dataSet: {histogram: [], keyList: [], totals: []},
+	loading: true,
+	publishDate: null,
+	timeRange: null,
+	timeSpanKey: null,
+	timeSpanOffset: 0,
+};
+
 const FALLBACK_DATA_SET_ITEM = {histogram: [], value: null};
 
-export const useChartState = ({publishDate, timeSpanKey}) => {
-	const {validAnalyticsConnection} = useContext(ConnectionContext);
+const ChartStateContext = createContext(INITIAL_STATE);
 
-	const [state, dispatch] = useReducer(reducer, {
-		dataSet: {histogram: [], keyList: [], totals: []},
-		loading: true,
+export const ChartStateContextProvider = ({
+	children,
+	publishDate,
+	timeRange,
+	timeSpanKey,
+}) => {
+	const stateAndDispatch = useReducer(reducer, {
+		...INITIAL_STATE,
 		publishDate,
+		timeRange,
 		timeSpanKey,
-		timeSpanOffset: 0,
 	});
 
-	const actions = {
-		addDataSetItems: (payload) =>
-			dispatch({
-				payload,
-				type: ADD_DATA_SET_ITEMS,
-				validAnalyticsConnection,
-			}),
-		changeTimeSpanKey: (payload) =>
-			dispatch({payload, type: CHANGE_TIME_SPAN_KEY}),
-		nextTimeSpan: () => dispatch({type: NEXT_TIME_SPAN}),
-		previousTimeSpan: () => dispatch({type: PREV_TIME_SPAN}),
-		setLoading: () => dispatch({type: SET_LOADING}),
-	};
+	return (
+		<ChartStateContext.Provider value={stateAndDispatch}>
+			{children}
+		</ChartStateContext.Provider>
+	);
+};
+
+export const useAddDataSetItems = () => {
+	const [, dispatch] = useContext(ChartStateContext);
+	const {validAnalyticsConnection} = useContext(ConnectionContext);
+
+	return (payload) =>
+		dispatch({
+			payload,
+			type: ADD_DATA_SET_ITEMS,
+			validAnalyticsConnection,
+		});
+};
+
+export const useChangeTimeSpanKey = () => {
+	const [, dispatch] = useContext(ChartStateContext);
+
+	return (payload) => dispatch({payload, type: CHANGE_TIME_SPAN_KEY});
+};
+
+export const useChartState = () => {
+	const [state] = useContext(ChartStateContext);
+
+	return state;
+};
+
+export function useDateTitle() {
+	const [state] = useContext(ChartStateContext);
+
+	const {histogram} = state.dataSet;
+	const {timeRange} = state;
+
+	return useMemo(() => {
+		if (histogram.length) {
+			const firstDateLabel = histogram[0].label;
+			const lastDateLabel = histogram[histogram.length - 1].label;
+
+			return {
+				firstDate: new Date(firstDateLabel),
+				lastDate: new Date(lastDateLabel),
+			};
+		}
+		else {
+			return {
+				firstDate: new Date(timeRange.startDate),
+				lastDate: new Date(timeRange.endDate),
+			};
+		}
+	}, [histogram, timeRange]);
+}
+
+export function useIsPreviousPeriodButtonDisabled() {
+	const [state] = useContext(ChartStateContext);
+
+	const {histogram} = state.dataSet;
+	const {publishDate} = state;
+
+	return useMemo(() => {
+		if (histogram.length) {
+			const firstDateLabel = histogram[0].label;
+
+			const firstDate = new Date(firstDateLabel);
+			const publishedDate = new Date(publishDate);
+
+			return firstDate < publishedDate;
+		}
+
+		return true;
+	}, [histogram, publishDate]);
```

Comment:
This is an example of a `useMemo` call that doesn't look worth it. The contents will be super fast to compute and create very little garbage (two `Date` objects), so I wouldn't expect this to have any measurable effect.

Reply:
This `useMemo` was used in Chart component with the date formatter call. I just move the code. So I can remove the `useMemo`, right?

Reply:
I'd think that you could.